### ISSUE TITLE
Explore different autodiff backends

### ISSUE THREAD

**Issue by fph** [JuliaDynamics/RigorousInvariantMeasures.jl#157]
I am opening this issue just to keep a log of my experiments with different differentiation backends; now that we have set up some basic regression tests we can experiment a little more with, e.g., `ForwardDiff` or `DualNumbers`.

**Comment by fph** [comment]
On the same machine, with TaylorSeries:
```
12×7 Matrix{Any}:
 "Experiment"          "Ass.time"   "max diam"   "mean diam"     "dfly time"    "dfly A"     "dfly B"
 "Lanford_Ulam"       1.16188      6.46594e-13  2.27241e-15     0.783834       0.666667     0.444444
 "Lanford3_Hat"       1.59309      6.50147e-13  1.28581e-14     0.854002       0.297026     0.946648
 "4x_perturbed_Ulam"  1.0123       2.84217e-13  3.26854e-15     0.834248       0.266761     0.399115
 "4x_perturbed_Hat"   1.10958      6.46427e-14  8.21917e-16     0.768065       0.479698     0.558407
 "PiecewiseLinear"    2.23771      7.10543e-15  2.60751e-17     3.20034        0.8          8.0
 "PiecewiseLinear2"   1.71824      2.91323e-13  2.75436e-15     3.53929        0.666667    17.3022
 "175"                1.14488      5.68434e-14  1.51067e-15     1.15096        0.588235    17.0
 "175_nonlinear"      2.51324      2.91323e-13  2.49242e-15     3.44973        0.666667    17.3022
 "Lorenz2"            2.86495      8.59757e-13  4.42146e-15    26.5351         0.955289   753.692
 "Lorenz3"            1.05957      1.30029e-12  2.27529e-14   103.662          0.922003    48.4306
 "BZ"                 1.96652      3.32179e-12  6.12226e-15     1.50772       Inf          Inf
Commit: 4603114; Machine: 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz; Date: 2023-02-22T19:33:34.290
```
and with ForwardDiff:
```
12×7 Matrix{Any}:
 "Experiment"          "Ass.time"   "max diam"   "mean diam"    "dfly time"    "dfly A"     "dfly B"
 "Lanford_Ulam"       1.58963      6.46594e-13  2.27241e-15    1.24194        0.666667     0.444444
 "Lanford3_Hat"       1.71725      6.50147e-13  1.28581e-14    1.34196        0.297026     0.946648
 "4x_perturbed_Ulam"  1.21725      2.84217e-13  3.26854e-15    1.15499        0.266761     0.399115
 "4x_perturbed_Hat"   1.2863       6.46427e-14  8.21917e-16    1.31455        0.479698     0.558407
 "PiecewiseLinear"    1.74392      7.10543e-15  2.60751e-17    3.94189        0.8          8.0
 "PiecewiseLinear2"   2.51959      2.91323e-13  2.75436e-15    4.76161        0.666667    17.3022
 "175"                1.11458      5.68434e-14  1.51067e-15    1.04064        0.588235    17.0
 "175_nonlinear"      1.86534      2.91323e-13  2.49242e-15    5.05475        0.666667    17.3022
 "Lorenz2"            3.71725      8.2423e-13   9.1299e-15    16.7341         0.955263   753.692
 "Lorenz3"            1.18933      6.82121e-13  1.78461e-14   64.6992         0.92071     48.4306
 "BZ"                 3.33372      3.1406e-12   5.92921e-15    2.6496        Inf          Inf
Commit: cd81bf9; Machine: 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz; Date: 2023-02-23T11:42:58.640.
```
So oddly enough assembling with ForwardDiff seems slower but the two long dfly's seem faster (and the final results for them slightly better).

This is for a basic configuration in which I allocate new `result` buffers every time (but I use `SVector`s for the Hessians); I do not think there are significant savings by being smarter since there should be no heap allocations anyway thanks to the SVectors.

**Comment by fph** [comment]
This suggests sticking with TaylorSeries, since the assembly time is massively longer than the dfly time in all our experiments, but it would be interesting to investigate why there is this difference. Maybe one is more efficient for first derivatives and the other for second derivatives?

**Comment by fph** [comment]
More data from `BenchmarkTools`; with TaylorSeries:
```
julia> f = Lorenz().branches[1].f; x = @interval(0.4)
[0.399999, 0.400001]

julia> @benchmark derivative($f, $x)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  72.360 μs …  8.848 ms  ┊ GC (min … max): 0.00% … 98.55%
 Time  (median):     77.775 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   78.520 μs ± 87.770 μs  ┊ GC (mean ± σ):  1.11% ±  0.99%

      ▃▆█▅▄           ▁▂▃▂▃▃▃▂▂▁                               
  ▁▂▄▇█████▇▄▃▃▃▅▆▆▇█████████████▆▅▄▃▃▂▂▂▁▁▁▁▁▁▁▁▂▂▁▂▂▂▂▂▂▁▁▁ ▃
  72.4 μs         Histogram: frequency by time        87.9 μs <

 Memory estimate: 22.81 KiB, allocs estimate: 503.

julia> @benchmark second_derivative($f, $x)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  72.511 μs …  9.293 ms  ┊ GC (min … max): 0.00% … 98.68%
 Time  (median):     78.067 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   78.834 μs ± 92.228 μs  ┊ GC (mean ± σ):  1.16% ±  0.99%

      ▃██▃              ▁▂▁▂▂▂▁                                
  ▁▂▄█████▇▄▃▃▃▃▄▅▅▆▇███████████▇▆▄▄▃▃▂▂▁▁▁▁▁▁▁▁▂▂▁▂▂▂▂▂▂▁▁▂▁ ▃
  72.5 μs         Histogram: frequency by time        88.5 μs <

 Memory estimate: 22.91 KiB, allocs estimate: 503.
```
With `ForwardDiff`:
```
julia> @benchmark derivative($f, $x)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  125.559 μs …  7.729 ms  ┊ GC (min … max): 0.00% … 97.70%
 Time  (median):     131.765 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   132.649 μs ± 76.141 μs  ┊ GC (mean ± σ):  0.57% ±  0.98%

      ▄▇█▆▄     ▂▄▆▇▆█▆▆▂                                       
  ▁▂▃███████▅▄▆███████████▆▄▄▃▂▂▂▂▂▂▂▂▂▂▂▂▂▁▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▃
  126 μs          Histogram: frequency by time          148 μs <

 Memory estimate: 22.91 KiB, allocs estimate: 524.

julia> @benchmark second_derivative($f, $x)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  252.033 μs …   7.684 ms  ┊ GC (min … max): 0.00% … 95.98%
 Time  (median):     263.793 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   271.767 μs ± 131.227 μs  ┊ GC (mean ± σ):  0.81% ±  1.66%

  ▆█▅▅▇▅▂▁▁▂▂▃▃▂▁                                               ▂
  ███████████████████▇▇▅▅▆▅▅▅▄▅▅▅▄▅▄▄▄▄▄▄▄▄▅▂▅▃▄▅▄▄▅▅▄▅▅▄▅▄▄▄▅▅ █
  252 μs        Histogram: log(frequency) by time        448 μs <

 Memory estimate: 45.41 KiB, allocs estimate: 1037.
```
`ForwardDiff` seems uniformly slower on these microbenchmarks; and particularly so for the second derivative. The mysteries for now are two for me:

1. Why is `dfly_inf_der` so much faster with `ForwardDiff`? Maybe there is some sneaky type instability?
2. How can TaylorSeries manage to compute two derivatives with essentially the same amount of time and allocations as one?

**Comment by fph** [comment]
A possible explanation for 1. is the difference in diameter: TaylorSeries produces derivatives with larger diameter:
```
julia> dy = derivative(f, x); diam(dy)
3.552713678800501e-15

julia> ddy = second_derivative(f, x); diam(ddy)
1.0658141036401503e-14
```
vs. ForwardDiff:
```
julia> dy = derivative(f, x); diam(dy)
8.881784197001252e-16

julia> ddy = second_derivative(f, x); diam(ddy)
5.329070518200751e-15
```

**Comment by fph** [comment]
And more insight into what happens:

* the truly expensive operation is `Interval^Real` inside Lorenz; IntervalArithmetic states that in order to obtain a tight inclusion it is implemented by allocating `BigFloat`s.
* `ForwardDiff` implements the derivative of `x^a` as `a * x^(a-1)`, so it requires two interval powers. `TaylorSeries` instead uses the recursion in https://juliadiff.org/TaylorSeries.jl/stable/api/#TaylorSeries.pow! , which essentially boils down to `a*(x^a)/x`, hence it uses only one interval power. This explains why it is faster, and why the cost of computing two derivatives is almost the same as the cost of computing one.
* The implementation also explains why `TaylorSeries` returns worse enclosures: computing $\frac{x^a}{x}$ causes overestimation.

**Comment by fph** [comment]
TL;DR: Switching to ForwardDiff is probably not a good idea. ForwardDiff is slower than TaylorSeries; on the other hand it is slightly more accurate, and hence it makes the minimization procedure in `dfly_inf_der` faster. On balance though I think we care more about speed in `preimages` and the assembler: the numerical error in the diameter usually is smaller than the discretization error.
